### PR TITLE
feat(firebase_auth): add custom sign in 

### PR DESCRIPTION
**Pull request by pr-Mais**

### CODE REVIEW

File: packages/firebase_auth/firebase_auth_dart/lib/src/id_token_result.dart [hunk 1]
```diff
@@ -51,7 +49,7 @@ class IdTokenResult {
   Map<String, dynamic> get toMap {
     return {
       'authTimestamp': _data['auth_time'],
-      'claims': _data['claims'],
+      'claims': _data,
```
Comment (Salakar) [inline packages/firebase_auth/firebase_auth_dart/lib/src/id_token_result.dart:52]: Is this right (and same for line 38)? Docs [here](https://firebase.google.com/docs/auth/admin/create-custom-tokens#create_custom_tokens_using_a_third-party_jwt_library) indicate it's 'claims' in data for custom ones: 
image

Comment (Salakar) [reply packages/firebase_auth/firebase_auth_dart/lib/src/id_token_result.dart:52]: https://github.com/invertase/flutterfire_desktop/pull/55/files#diff-2d6bd4626abf4efad497dd814a49a20957370efbc18fd2989ee63763b95c4c00R91

this also indicates it should be _data['claims']

File: packages/firebase_auth/firebase_auth_dart/lib/src/providers/enums.dart [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      case ProviderId.phone:
+        return 'phone';
+      case ProviderId.facebook:
+        return 'facebook.com';
+      case ProviderId.github:
+        return 'github.com';
+      case ProviderId.twitter:
+        return 'twitter.com';
+      case ProviderId.custom:
+        return 'custom';
```
Comment (Salakar) [inline packages/firebase_auth/firebase_auth_dart/lib/src/providers/enums.dart:29]: Should this be 'firebase'? Firebase JS SDK seems to use 'firebase' based on those tests I linked earlier on chat 🤔

File: packages/firebase_auth/firebase_auth_dart/lib/src/id_token_result.dart (hunk 1 above)
Comment (pr-Mais) [reply packages/firebase_auth/firebase_auth_dart/lib/src/id_token_result.dart:52]: That's what I tried actually, I'm not sure if it's an issue with the way I'm testing but decoding the `idToken` I get from sign in in the emulator returns the claims as fields, e.g.:
```
// this is decoded into the custom token
{
  'uid': 'some-uid'
  'claims': { 'role': 'admin' }
}
// this is what the decoded idToken looks like after sign in
{
  'role': 'admin',
  // other reserved fields 
}
```

File: packages/firebase_auth/firebase_auth_dart/lib/src/providers/enums.dart (hunk 2 above)
Comment (pr-Mais) [reply packages/firebase_auth/firebase_auth_dart/lib/src/providers/enums.dart:29]: the type returned and decoded in the idToken is `custom`, see here: https://github.com/firebase/firebase-js-sdk/blob/0b3ca78eb97ce328b7db4ad5bb11f254e5de6a9a/packages/auth/src/model/enums.ts#L26 & https://github.com/firebase/firebase-js-sdk/blob/0b3ca78eb97ce328b7db4ad5bb11f254e5de6a9a/packages/auth/src/api/authentication/custom_token.test.ts

Comment (Salakar) [reply packages/firebase_auth/firebase_auth_dart/lib/src/providers/enums.dart:29]: thanks for confirming!

File: packages/firebase_auth/firebase_auth_dart/lib/src/id_token_result.dart (hunk 1 above)
Comment (pr-Mais) [reply packages/firebase_auth/firebase_auth_dart/lib/src/id_token_result.dart:52]: The JS SDK does the same internally, so I suppose the docs explain the returned `IdTokenResult` from the SDKs not the raw request
https://github.com/firebase/firebase-js-sdk/blob/cdada6c68f9740d13dd6674bcb658e28e68253b6/packages/auth/src/core/user/id_token_result.ts

**Pull request merged**